feat(docs): add wallet tutorial (#20111)
AztecProtocol/aztec-packages: docs/docs-developers/docs/tutorials/js_tutorials/webapp/02-network-and-wallet.md

---
title: "2. Network & Wallet"
sidebar_position: 2
description: "Connect to Aztec using an embedded wallet for local dev or the wallet SDK for devnet"
---

# Network & Wallet

This section covers connecting your webapp to the Aztec network and setting up a wallet. You'll create three source files (`config.ts`, `embedded-wallet.ts`, `wallet-connection.ts`) plus shared fee utilities and UI components.

## Key concepts

Before writing any code, it helps to understand two pieces of infrastructure that every Aztec app relies on:

> **PXE (Private eXecution Environment)** is a client-side runtime that runs in the browser as WASM. It stores your private notes, manages your encryption keys, executes private functions, and generates zero-knowledge proofs. Because PXE runs locally, your private data never leaves the browser. Every Aztec app needs a PXE — either one it creates itself or one provided by a wallet extension.

Reviewer: Add links to relevant docs pages.

> **Aztec node** is the server-side component that maintains the network's public state and sequences transactions into blocks. Your PXE connects to a node (a local sandbox during development, or a remote devnet/mainnet node in production) to sync state and submit transactions. The node never sees your private data — it only receives proofs and encrypted outputs.

Reviewer: ```suggestion
**Aztec node** is the server-side component that maintains the network's public state and sequences transactions into blocks. Your PXE connects to a node (a local network during development, or a remote devnet/mainnet node in production) to sync state and submit transactions. The node never sees your private data — it only receives proofs and encrypted outputs.
```

The relationship is straightforward: PXE handles everything private (notes, keys, proofs), the node handles everything public (state, blocks, sequencing), and they communicate over a standard RPC interface.

## Network configuration

Create `src/config.ts` to determine which Aztec node to connect to and to provide a helper for creating an in-browser PXE:

#include_code config /docs/examples/webapp-tutorial/src/config.ts typescript

`createLocalPXE` sets up PXE in three steps:

1. `createAztecNodeClient` — opens an RPC connection to the Aztec node so PXE can sync public state and submit transactions.
2. `getPXEConfig` + `getL1ContractAddresses` — fetches protocol configuration from the node, including L1 contract addresses and network parameters that PXE needs to construct valid proofs.
3. `createPXE` — starts a full PXE instance in the browser. From this point on, all private execution happens locally.

## Wallet modes

Aztec supports two wallet modes, and you'll implement both:

- **Embedded wallet** — your app creates PXE and manages accounts directly. Best for local development with the sandbox.

> - **Wallet SDK** — your app connects to an external browser extension that owns PXE and accounts. Required for devnet and production.

Reviewer: "Required for devnet and production."

Is this true? I can imagine use cases where embedded wallets are useful for new users that don't have an Aztec wallet yet, although I haven't totally thought through what the user journey would be

Both modes produce the same `Wallet` interface, so the rest of your app doesn't need to know which one is in use.

## Embedded wallet (local development)

For local development, you'll create an `EmbeddedWallet` class that extends `BaseWallet` from the Aztec wallet SDK. `BaseWallet` provides account abstraction, transaction signing, and fee handling — your subclass just needs to supply initialization logic, account management, and a fee payment strategy.

Create `src/embedded-wallet.ts`:

### Initialization

> #include_code initialize /docs/examples/webapp-tutorial/src/embedded-wallet.ts typescript

Reviewer: I paste in the code and get syntax errors. I think it'd be nice to start with the imports and class definition and build up the file without errors

Reviewer: I don't think the imports section is anywhere in this file

This is similar to `createLocalPXE` from `config.ts`, but additionally registers the SponsoredFPC contract with PXE so that fee payment works out of the box.

### Connecting a test account

> The local sandbox ships with pre-deployed test accounts. These are Schnorr-signature accounts that are already registered and funded on the sandbox, so you can use them immediately without deploying a new account contract. You select one by index (0, 1, 2, etc.).

Reviewer: ```suggestion
The local sandbox ships with 3 pre-deployed test accounts. These are Schnorr-signature accounts that are already registered and funded on the sandbox, so you can use them immediately without deploying a new account contract. You select one by index (0, 1, 2, etc.).
```

#include_code connect-test-account /docs/examples/webapp-tutorial/src/embedded-wallet.ts typescript

### Fee payment

Every Aztec transaction must pay a fee (similar to gas on Ethereum). Rather than requiring users to hold fee tokens during development, the embedded wallet overrides `completeFeeOptions` to inject SponsoredFPC as the default fee payer for every transaction. Callers never need to pass fee options manually.

> #include_code fee-options /docs/examples/webapp-tutorial/src/embedded-wallet.ts typescript

Reviewer: not sure that this is totally necessary. I think just the text is fine, I dont need to see the implementation

### Full class

The complete `EmbeddedWallet` puts these pieces together:

> #include_code embedded-wallet-class /docs/examples/webapp-tutorial/src/embedded-wallet.ts typescript

Reviewer: this is missing the imports, so i can't just copy-paste it

## Wallet SDK (devnet / production)

> On devnet and production, users connect via a browser extension wallet (like MetaMask on Ethereum). The wallet extension owns the PXE, manages keys, and signs transactions. Your app communicates with it through the **wallet SDK**, which handles discovery, secure channel setup, and verification.

Reviewer: ```suggestion
On devnet and production, users connect via a browser extension wallet (like MetaMask on Ethereum). The wallet extension owns the PXE, manages keys, and generated proofs. Your app communicates with it through the **wallet SDK**, which handles discovery, secure channel setup, and verification.
```

### Testing with the test wallet extension

> The tutorial includes a minimal test wallet extension in `test-extension/` for testing the wallet SDK flow without installing a real wallet. This extension auto-approves all connection requests and returns a mock account — useful for development and E2E testing.

Reviewer: link to test-extension

> To use it:

Reviewer: These instructions aren't super clear. As a reader of the tutorial idk where the webapp-tutorial dir is.

Reviewer: maybe we will want to mirror the examples/webapp-tutorial folder outside of the monorepo?

1. Build the extension (from the `webapp-tutorial` directory):
   ```bash
   node esbuild.extension.mjs
   ```

2. Load it in Chrome:
   - Navigate to `chrome://extensions/`
   - Enable "Developer mode" (toggle in top-right)
   - Click "Load unpacked"
   - Select the `test-extension/` folder

3. After making changes to the extension source, rebuild and click the refresh icon in `chrome://extensions/`

> The test extension responds to `getAccounts()` with a hardcoded test account. It won't sign real transactions, but it's enough to verify your wallet connection UI works correctly before testing with a real wallet.

Reviewer: Is building the test extension a requirement before continuing? its not clear

> Create `src/wallet-connection.ts`:

Reviewer: This feels like part of a new step. Move this line below the step one heading.

### Step 1: Discover available wallets

Your app needs to find which wallet extensions the user has installed. The SDK does this through a `window.postMessage`-based discovery protocol: your app broadcasts a discovery request, and any installed wallet extension responds with its provider info (name, icon, supported chain).

The `discoverWallets` function starts this process and calls your `onUpdate` callback each time a new wallet extension responds. You use the resulting list to show users a "pick your wallet" UI.

> #include_code discover-wallets /docs/examples/webapp-tutorial/src/wallet-connection.ts typescript

Reviewer: missing imports and types above this

Reviewer: also not sure that i care about all of the implementation details here. Wdyt about creating a button that copies all of the specified source file contents without having to display it, so I can just click a button that says "Copy wallet-connection.ts code", and drop it in my editor and inspect it if i want? I think the text you have is a good high-level explainer, but we don't go deep into the code in the text and I can imagine people want to get to something working with less reading. They can read the code in their IDE if they want
